Date: Fri, 29 May 2020 17:54:57 +0000
From: mselva006c <mani_selvaraj@comcast.com>
Subject: [PATCH] RDK-28776: force reboot on systemd freeze
          Added a force reboot to ensure the box reboot on systemd freeze
Source: COMCAST 2eb0a0ceaf86e678cd2c22438da5fb493fd17eba Mon Sep 17 00:00:00 2001
Signed-off-by: mselva006c <mani_selvaraj@comcast.com>
---
 src/core/main.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/core/main.c b/src/core/main.c
index c24b696b16..12265f1ec6 100644
--- a/src/core/main.c
+++ b/src/core/main.c
@@ -152,6 +152,8 @@ static int parse_configuration(const struct rlimit *saved_rlimit_nofile,
 
 _noreturn_ static void freeze_or_exit_or_reboot(void) {
 
+        // create a FILE typed pointer
+        FILE *file_pointer;
         /* If we are running in a container, let's prefer exiting, after all we can propagate an exit code to
          * the container manager, and thus inform it that something went wrong. */
         if (detect_container() > 0) {
@@ -169,6 +171,19 @@ _noreturn_ static void freeze_or_exit_or_reboot(void) {
         }
 
         log_emergency("Freezing execution.");
+        log_info("Systemd went into freeze Mode, we will do a force reboot on next reboot call");
+        // open the file "name_of_file.txt" for writing
+        file_pointer = fopen("/tmp/systemd_freeze_reboot_on", "w");
+        if ( file_pointer != NULL )
+        {
+          // Write to the file
+          fprintf(file_pointer, "REBOOT");
+
+          // Close the file
+          fclose(file_pointer);
+        }else {
+           log_error("Failed to open reboot flag file");
+        }
         freeze();
 }
 
